### DISCUSSION TITLE
[Possible Bug] Assets registered in Panel Provider are not respecting X-Forwarded-Host Header

### DISCUSSION THREAD

**Discussion by nilskretschmer** [filamentphp/filament#14308] [Help]
### Package

Panel builder

### Package Version

v3.2.101

### Issue?
I would open up an issue for this but I am not sure how to provide a reproduction repository as this would require external resources (load balancer/proxy) to be included making "just" a reproduction repository difficult. Alternatively I could provide a minimal App on the internet behind a proxy to showcase the problem. But that is also not allowed for Filament issues as those issues get closed immediately. Maybe the Filament-Team could give me a hint?

### How can we help you?

I have a very strange issue using Filament in an enterprise/production environment. I defined some additional resources in my PanelProvider:

    public function panel(Panel $panel): Panel
    {
        return $panel
            ->default()
            ->font('Benton', url: Vite::asset('resources/css/fonts.css'), provider: LocalFontProvider::class)
            ->viteTheme('resources/css/filament/admin/theme.css')
            ->brandLogo(asset('img/logo.svg'))
            ->darkModeBrandLogo(asset('img/logo-dark.svg'));
    }
    
The app is accessed using a proxy and an external URL that differs from the internal URL that is being used from an internal network. All assets are retrieved using HTTPS. I also set the trusted proxies (https://laravel.com/docs/11.x/requests#configuring-trusted-proxies) and this works as expected. The only issue is that the assets defined in the Filament Panel Provider are being served using HTTP(S) using the INTERNAL domain/URL (see attached). As you can see all assets - except for those defined in the panel provider - are correctly served over HTTPS and under the external domain the app is being accessed with.

So is Filament doing something different here by returning asset URLs using the internal App URL? 

![grafik](https://github.com/user-attachments/assets/5c7e3b99-54f4-4581-8882-fb94e7efe402)

If I explicitly set the host header to the `X-Forwarded-Host`-Header in the Panel Provider like:
```
public function panel(Panel $panel): Panel
{
  if (request()->hasHeader('X-Forwarded-Host') {
      request()->headers->set('Host', request()->header('X-Forwarded-Host'));
  }
}
```

The assets are now correctly served by Laravel/Filament from the EXTERNAL Url (host):
![grafik](https://github.com/user-attachments/assets/8f2af4e6-2be5-4f61-8d1f-fa447e723ac9)

I also double-checked that the `X-Forwarded-Host`-Header contains the expected EXTERNAL Url which is correct.

This seems to be very odd. Laravel respects the `X-Forwarded-Host`-Header per default when running behind a loadbalancer if `TrustProxies.php` is configured correctly. I didn't ever experience issues with assets but only using Filament in this case... Any ideas?
This is my `app/Http/Middleware/TrustProxies.php`
```
<?php
 
namespace App\Http\Middleware;
 
use Illuminate\Http\Middleware\TrustProxies as Middleware;
use Illuminate\Http\Request;
 
class TrustProxies extends Middleware
{
    /**
     * The trusted proxies for this application.
     *
     * @var string|array
     */
    protected $proxies = '*';
 
    /**
     * The headers that should be used to detect proxies.
     *
     * @var int
     */
    protected $headers = Request::HEADER_X_FORWARDED_FOR | Request::HEADER_X_FORWARDED_HOST | Request::HEADER_X_FORWARDED_PORT | Request::HEADER_X_FORWARDED_PROTO;
}
```

**Comment by patrickcarlohickman** [comment]
@nilskretschmer 

We just ran into this trying to set the favicon. We actually found this issue because of this. The application is behind a load balancer, so, even though we have trusted proxies setup, the `Vite::asset()` call was resolving to `http` instead of `https`.

The issue is that the code in the panel registration is run when the PanelRegistry is resolved, which is before trusted proxies takes effect. Therefore, if the call to generate the asset url is executed inside the `panel()` method, it won't know anything about the trusted headers. You need a way to delay the call to generate the asset url.

With the `favicon()` method, it accepts a callback that will be evaluated when the favicon is rendered. This allowed us to get around this issue. However, not all the methods accept a callback. For example, the `font()` method only accepts strings.

Looking into this a little more, I saw there was a `bootUsing()` method that lets you define boot callback methods. This should get you where you want to be. Anything that needs an generated asset url, you should throw inside of a boot callback. Modifying your provided code, this would look like:

```
public function panel(Panel $panel): Panel
{
    return $panel
        ->default()
        ->viteTheme('resources/css/filament/admin/theme.css') // hardcoded relative url; not generated
        ->bootUsing(function (Panel $panel) {
            $panel
                ->font('Benton', url: Vite::asset('resources/css/fonts.css'), provider: LocalFontProvider::class)
                ->brandLogo(asset('img/logo.svg'))
                ->darkModeBrandLogo(asset('img/logo-dark.svg'));
        });
}
```

Hope this helps.

Thanks,
Patrick
